### DISCUSSION TITLE
Desired TP+ Functionality

### DISCUSSION THREAD

**Discussion by kobbled** [kobbled/tp_plus#18] [Ideas]
Any desired functionality for TP+ that has not been implemented, and is not an immediate issue for your usage of TP+, please post here.

**Comment by kobbled** [comment]
Need Independent and Simultaneous motion group instructions. See https://github.com/kobbled/tp_plus/issues/6#issuecomment-914583227.

**Comment by dickvand** [comment]
Add the distance_before (DB) option to the motion command, to be able to start a command or write an output on a fixed distance before reaching the end point of the movement.

**Reply by kobbled** [reply]
I don't know why I didn't add that before, it wasn't hard to add in. see [2243f29](https://github.com/kobbled/tp_plus/commit/2243f29b58872370fc7fdccb204202716353a750). I still haven't put in point logic though, hopefully that is not a big deal.

This should work
```
point1 := P[1]
pin1 := DI[1]

use_uframe 0
use_utool 1

point1.joints -> [0,0,0,0,0,0]

linear_move.to(point1).at(100, 'mm/s').term(0).distance_before(100, foo())
linear_move.to(point1).at(100, 'mm/s').term(0).distance_before(100, turn_on(pin1))
```

**Comment by dickvand** [comment]
Hi Matt,

Thanks for your action!
We would like to write a value to a group output at the distance before. Can you add that functionality as well?

Kind regards,
Dick

Van: Matt Dewar ***@***.***>
Verzonden: zaterdag 9 april 2022 05:30
Aan: kobbled/tp_plus ***@***.***>
CC: dickvand ***@***.***>; Comment ***@***.***>
Onderwerp: Re: [kobbled/tp_plus] Desired TP+ Functionality (Discussion #18)


I don't know why I didn't add that before, it wasn't hard to add in. see 2243f29<https://github.com/kobbled/tp_plus/commit/2243f29b58872370fc7fdccb204202716353a750>. I still haven't put in point logic though, hopefully that is not a big deal.

This should work

point1 := P[1]

pin1 := DI[1]



use_uframe 0

use_utool 1



point1.joints -> [0,0,0,0,0,0]



linear_move.to(point1).at(100, 'mm/s').term(0).distance_before(100, foo())

linear_move.to(point1).at(100, 'mm/s').term(0).distance_before(100, turn_on(pin1))

**Reply by kobbled** [reply]
whoops sorry. It should work now. I used the wrong parser rule.

```
p := P[1]
d := DI[1]
g := GO[1]
p := P[1]
linear_move.to(p).at(2000, 'mm/s').term(0).distance_before(100, d=on)
linear_move.to(p).at(2000, 'mm/s').term(0).distance_before(100, g=50)
```

**Comment by kobbled** [comment]
Next feature to add will be local variables. These will be handled by keyword mappings `local`, `localpr` in an env file, or at the top of a program:

```
local := R[50..100]
localpr := PR[5..10]
```

Then in a program, namespace, or function, variables can be declared

```
def foo(arg1, arg2) : numreg
   v1 := LR[1]
   v2 := LR[2]
   pr1 := LPR[1]
  
   v3 := LR[3]
   v3 = v1*v2

  Pos::clrpr(&pr1)
  Pos::setxyz(0,0,0,0,0,v3)

  linear_move.to(pr1).at(2000, 'mm/s').term(-1)  

end
```

The local variables will be set to an available physical register from the `local`, or `localpr` groups. The state of the physical registers will be tracked via a virtual stack when interpreting the file. During compilation the actual register will be replaced by the local register. Once out of scope a default value will be set in the register. With a virtual stack created, scoping will need to be tracked, as well as the ability to move through the program tree during compilation. Because of this all function and namespaces will need to be explicitly linked using `imports`. A method for declaring karel functions in TP+ will need to be implemented because of this:

```
karel namespace Pos
   karel def clrpr(p) end
   karel def setxyz(x,y,z,w,p,r) : posreg end
end
```

This feature will enable expressions to be able to be set in CALL arguments, i.e.:
```
foo((i+j), Mth::arcang(6, 50))
```

A feature issue has been raised to follow this in #3
